lib/Sema/CSSyntacticElement.cpp

```diff
@@ -615,9 +616,10 @@ class SyntacticElementConstraintGenerator
       return;
     }
 
-    // Convert the contextual type to the pattern, which establishes the
-    // bindings.
-    cs.addConstraint(ConstraintKind::Conversion, context.getType(), patternType,
+    // Require a conversion from the pattern type (which is always r-value) to
+    // the r-value contextual type, which establishes the bindings.
+    const auto contextualTy = context.getType()->getRValueType();
+    cs.addConstraint(ConstraintKind::Conversion, patternType, contextualTy,
```

Comment:
This doesn't seem correct to me, conversions should propagate _into_ patterns, not out of them (the opposite of expressions).  For example, for the following:

```swift
func foo(_ x: Any) {
  if case let y: Int = x {}
}
```

we don't want to setup a conversion constraint from the TypedPattern type `Int` to the expression type `Any`, as that would succeed, but that's incorrect as we can't pattern match `Any` as `Int` (at least not without a cast).

I suspect that you changed this to allow the optional promotion to work, but you'll instead want to change `getTypeForPattern` to allow tuple patterns to differ in optionality.

Reply:
Took me some time to digest but I see the point now. What about the following? This is currently rejected outside of closures but accepted in closures presumably due to the direction of the constraint. The reverse matching is —expectedly, I’d say — accepted in either context.

```swift
switch (0, 0) {
  case (x: 0, y: 0): break
  default: break
}
```

Reply:
Heh interesting, we accept these too in any context:

```swift
let (x: x, y: y) = (0,0)
if case (x: 0, y: 0) = (0,0) {}
for case (x: 0, y: 0) in [(0,0)] {}
```

As such I'm inclined to say we should just allow it for consistency, even if it's weird. We have a general rule that tuple conversions can add or remove labels, as long as they don't rename them, and this seems to fit with that. Changing that established rule would probably warrant some Swift evolution discussion.

Any thoughts @xedin?

Reply:
I agree with your sentiment @hamishknight and the original comment as well. I think it makes sense to allow this in general.

Reply:
> you'll instead want to change `getTypeForPattern` to allow tuple patterns to differ in optionality.

@hamishknight Were you anticipating a conversion constraint to work for this purpose?

Reply:
@AnthonyLatsis Given you only want to allow optional promotion, a conversion isn't quite right as it'll allow other conversions too. I think the best way to handle this would probably be to add a new constraint that waits until the optionality of the RHS is resolved, and then unwraps and binds it to the LHS.

Reply:
It's like `OptionalObject` in reverse :)

Reply:
Sounds fun, I reached a pretty similar conclusion. A new constraint it is then; thanks!

> It's like `OptionalObject` in reverse :)

The difference in optionality depth is `>= 0` for this one though.